subsys/net/ip/tcp.c

```diff
@@ -2639,124 +2666,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
-				/* Cases:
-				 * - Data already received earlier: len > 0 , new_len <= 0
-				 * - Partially new data len > 0, new_len > 0
-				 * - Out of order data len > 0, new_len > 0, ignore the data in
-				 *   between
+			/* Compute if there is new data after our close */
+			if (tcp_compute_new_length(conn, th, len) > 0) {
+				/* We do not implement half closed sockets, therefore
+				 * cannot accept new data in after sending our FIN, as
+				 * we are in sequence can send a reset now.
 				 */
-				if (new_len > 0) {
-					conn_ack(conn, + new_len);
-				}
-			}
+				net_stats_update_tcp_seg_drop(conn->iface);
 
-			if (FL(&fl, ==, FIN, th_seq(th) == conn->ack) ||
-				   FL(&fl, ==, FIN | ACK, th_seq(th) == conn->ack) ||
-				   FL(&fl, ==, FIN | PSH | ACK,
-				      th_seq(th) == conn->ack)) {
-				/* Received FIN on FIN_WAIT_2, so cancel the timer */
-				k_work_cancel_delayable(&conn->fin_timer);
+				next = tcp_enter_time_wait(conn);
 
+				tcp_out(conn, RST);
+				break;
+			}
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check the sequence number of the FIN flag with tha ack
+			 */
+			if (FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) {
 				conn_ack(conn, + 1);
-				tcp_out(conn, ACK);
-				next = TCP_TIME_WAIT;
-				verdict = NET_OK;
-			} else if (len > 0) {
-				/* Send out a duplicate ACK */
-				tcp_out(conn,  ACK);
+				NET_DBG("FIN received, going to TIME WAIT");
+
+				next = tcp_enter_time_wait(conn);
+
 				verdict = NET_OK;
+				tcp_out(conn, ACK);
+			} else {
+				if (len > 0) {
+					/* Send out a duplicate ACK */
+					tcp_out(conn, ACK);
+					verdict = NET_OK;
+				}
 			}
 		}
 		break;
 	case TCP_CLOSING:
-		if (th && FL(&fl, ==, ACK, th_seq(th) == conn->ack)) {
-			tcp_send_timer_cancel(conn);
-			next = TCP_TIME_WAIT;
-			verdict = NET_OK;
+		if (th) {
+			/*
+			 * Closing:
+			 * Our FIN has to be acknowledged
+			 * -   -> TCP_CLOSING
+			 * ACK -> TCP_TIME_WAIT
+			 */
+			int32_t new_len = tcp_compute_new_length(conn, th, len);
+			/* Count with -1 as the FIN flag has to be subtracted */
+			if ((new_len > -1) && (len > 0)) {
+				/* This should not happen here, as no data can be send after
+				 * the FIN flag has been send.
+				 */
+				NET_ERR("conn: %p, new bytes %u during CLOSING state "
+					"sending reset", conn, new_len);
+				net_stats_update_tcp_seg_drop(conn->iface);
+
+				next = tcp_enter_time_wait(conn);
+
+				tcp_out(conn, RST);
+				break;
+			}
+
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				NET_DBG("FIN acknowledged, going to TIME WAIT "
+					"state seq %u, ack %u"
+					, conn->seq, conn->ack);
+
+				next = tcp_enter_time_wait(conn);
+
+				verdict = NET_OK;
+			}
+
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check with the ack
+			 */
+			if ((FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) || (len > 0)) {
+				if (next == TCP_TIME_WAIT) {
```

Comment:
Maybe introduce `fin_acked` helper variable in this state as well, as suggested above?

Reply:
Done